### PR TITLE
Java harness

### PR DESCRIPTION
**Pull request by cromwellian**
Portable interfaces to be shared with Android, while allowing rapid development in Chrome.

Most of these are skeleton examples, to be filled in later. See the README.md

---
This change is [Reviewable](https://reviewable.io/reviews/polymerlabs/arcs/3152)

### CODE REVIEW

File: src/javaharness/java/arcs/api/PortableJson.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
+package arcs.api;
+
+/**
+ * Portable representation of JSON structure.
+ */
+public interface PortableJson {
```
Comment (shaper) [inline src/javaharness/java/arcs/api/PortableJson.java]: Is this and the parser new/original code or brought in from somewhere existing?

File: src/javaharness/java/arcs/nativeparticles/EchoParticle.java [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package arcs.nativeparticles;
```
Comment (shaper) [inline src/javaharness/java/arcs/nativeparticles/EchoParticle.java]: We should rename directories/files with "native" in the title in a later PR. Something short but still meaningful. To bikeshed further.

File: src/javaharness/java/arcs/webimpl/WebHarnessController.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
+package arcs.webimpl;
+
+import arcs.api.ArcsEnvironment;
+import arcs.api.DeviceClient;
+import elemental2.dom.*;
```
Comment (shaper) [inline src/javaharness/java/arcs/webimpl/WebHarnessController.java]: What is this, and is it a new dependency?

File: src/javaharness/README.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+Native particles will reside and be registered in arcs/nativeparticles
+
+Note there are symbolic links between pipes-shell and
+the outer shells/pipes-shell directory, that means the pipes-shell 
+must be build first.
+
+## Running
+
+First, ensure xcode + command lines tools are installed.
```
Comment (shaper) [inline src/javaharness/README.md]: xcode?

File: src/javaharness/README.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+
+Code in arcs/api is platform independent and should have no dependencies on Android API runtime.
+
+Code in arcs/webimpl is platform dependent and expected to be reimplemented in arcs/androidimpl
+
+Native particles will reside and be registered in arcs/nativeparticles
+
+Note there are symbolic links between pipes-shell and
+the outer shells/pipes-shell directory, that means the pipes-shell 
+must be build first.
```
Comment (shaper) [inline src/javaharness/README.md]: Add command/instruction on how to build pipes-shell, or link to existing build info if exists in another .md already?

File: src/javaharness/.bazelrc [hunk 6]
```diff
@@ -0,0 +1,11 @@
+# Recommended bazel settings for working with J2CL.
+# You can copy this into root of your workspace.
+
+build --incompatible_depset_union=false
+#build --symlink_prefix=/
```
Comment (shaper) [inline src/javaharness/.bazelrc]: Do we need this?

File: src/javaharness/README.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+## Running
```
Comment (shaper) [inline src/javaharness/README.md]: What version of Java is required and how to ensure as part of install/setup/build that the developer has it?

File: src/javaharness/java/arcs/api/ArcsEnvironment.java [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+public interface ArcsEnvironment {
+
+  /** Called by ArcsEnvironment when new suggestions are available from Arcs. */
+  interface SuggestionListener {
+    void onSuggestion(String arcId, Collection<String> suggestions);
+  }
+
+  /** Called by ArcsEnvironment when the Arcs runtime is ready. */
+  interface ReadyListener {
+    void onReady(List<String> autofillTypes);
```
Comment (shaper) [inline src/javaharness/java/arcs/api/ArcsEnvironment.java]: For generality removing the text 'autofill' from the PR feels worthwhile, use alternate name or remove methods/etc. as feasible? Open to thoughts and could be another PR.

**Comment by shaper** [comment]
Just to comment on quickly-to-follow work, IMO we should land this (after edits), @yuangu-google should move us to pipes-shell-2, and we should immediately update this to be compatible with that. If we encounter notable work as part of that let's discuss. @sjmiles @mmandlis

File: src/javaharness/java/arcs/webimpl/WebHarnessController.java (hunk 3 above)
Comment (cromwellian) [reply src/javaharness/java/arcs/webimpl/WebHarnessController.java]: Elemental is the glue between Java and the WebAPIs for GWT and J2CL. Think of it as the Java equivalent of Typescript's .d.ts files. It has zero code-size and run-time overhead.

File: src/javaharness/README.md (hunk 4 above)
Comment (cromwellian) [reply src/javaharness/README.md]: Yes, bazel requires it because it builds some stuff from source and some of the dependencies have c code that needs compilation. It's exactly like blaze, so for example, the protobuf compiler used by bazel itself is written in C++.

File: src/javaharness/java/arcs/api/PortableJson.java (hunk 1 above)
Comment (cromwellian) [reply src/javaharness/java/arcs/api/PortableJson.java]: This is a new interface I provide. It will allow the harness code to parse JSON and not have to directly reference either Android's org.json package or JS APIs.  It's a very thin utility class that is backed by platform dependent parser. This allows the api/ dir to contain platform independent code.

File: src/javaharness/.bazelrc (hunk 6 above)
Comment (cromwellian) [reply src/javaharness/.bazelrc]: Yes, bazel is in the process of a great migration to new skylark rules and some of them haven't been migrated yet.

File: src/javaharness/README.md (hunk 7 above)
Comment (cromwellian) [reply src/javaharness/README.md]: Java8 or above. Most MacOS developer devices already have  Java I think, but I guess we can add a script to download it.

File: src/javaharness/java/arcs/api/ArcsEnvironment.java (hunk 8 above)
Comment (cromwellian) [reply src/javaharness/java/arcs/api/ArcsEnvironment.java]: I'll remove it in a follow up.

File: src/javaharness/README.md (hunk 5 above)
Comment (cromwellian) [reply src/javaharness/README.md]: There's an explanation of how to do it by running deploy.sh a few lines below this.
